src/coreclr/tools/Common/Compiler/GenericCycleDetection/ModuleCycleInfo.cs

```diff
@@ -191,7 +209,20 @@ private bool FormsCycle(TypeSystemEntity entity, out ModuleCycleInfo cycleInfo)
                 EcmaModule ownerModule = (entity as EcmaType)?.EcmaModule ?? (entity as EcmaMethod)?.Module;
                 if (ownerModule != null)
                 {
-                    cycleInfo = _hashtable.GetOrCreateValue(ownerModule);
+                    if (!_hashtable.TryGetValue(ownerModule, out cycleInfo))
+                    {
+                        if (ownerModule.MetadataReader.GetTableRowCount(TableIndex.TypeSpec) < MinimumTypeSpecCountToAnalyzeSingleThreaded)
+                        {
+                            cycleInfo = _hashtable.GetOrCreateValue(ownerModule);
+                        }
+                        else
+                        {
+                            lock (ownerModule)
```

Comment:
We'd need to remember that nobody else should lock on EcmaModules so that we don't get an accidental deadlock somewhere.

I don't have strong opinions about this - the compilation will finish even without doing this and whether it take 2 minutes or 8 minutes doesn't make a huge difference - it's already orders of magnitude longer than usual - the assembly is simply AOT hostile and it will cause trouble for any AOT we have in .NET (NativeAOT, Mono AOT, crossgen2). I would be fine won't fixing that part.